src/blockrequest.cpp

```diff
@@ -0,0 +1,128 @@
+    requestedUpToSize = 0;
+    processedUpToSize = 0;
+}
+
+CBlockRequest::~CBlockRequest()
+{
+    LogPrint("net", "Deallocating CBlockRequest\n");
+}
+
+void CBlockRequest::processWithPossibleBlock(const CBlock* pblock, CBlockIndex *pindex)
+{
+    // don't process anything if request was cancled
+    if (this->fCancled)
+        return;
+
+    int MAX_PROCESS = 5;
+    int loop_processed = 0;
+    for (unsigned int i = this->processedUpToSize; i < this->vBlocksToDownload.size() ; i++) {
+        CBlockIndex *pindexRequest = this->vBlocksToDownload[i];
+        CBlock loadBlock;
+        const CBlock *currentBlock = &loadBlock;
+
+        // if a block has been passed, check if is the next item in the sequence
+        if (pindex && pblock && pindex == pindexRequest)
+            currentBlock = pblock;
+        else if (pindexRequest->nStatus & BLOCK_HAVE_DATA) {
+            if (!ReadBlockFromDisk(loadBlock, pindexRequest, Params().GetConsensus()))
+                throw std::runtime_error(std::string(__func__) + "Can't read block from disk");
+        } else {
+            break;
+        }
+
+        // fire signal with txns
+        int cnt = 0;
+        BOOST_FOREACH(const CTransaction &tx, currentBlock->vtx) {
+            GetMainSignals().SyncTransaction(tx, pindexRequest, cnt, false);
+            cnt++;
+        }
+        this->processedUpToSize++;
+
+        // log some info
+        LogPrint("net", "BlockRequest: proccessed up to %ld of total requested %ld blocks\n", this->processedUpToSize, this->vBlocksToDownload.size());
+
+        if (progressCallback)
+            if (!progressCallback(shared_from_this(), pindexRequest))
+                this->cancel();
+
+        // release global block request pointer if request has been completed
+        if (this->processedUpToSize == this->vBlocksToDownload.size())
+            currentBlockRequest = nullptr;
+
+        if (loop_processed >= MAX_PROCESS)
+            break;
+        loop_processed++;
+    }
+}
+
+void CBlockRequest::cancel()
+{
+    fCancled = true;
+    if (currentBlockRequest.get() == this) {
+        // release shared pointer
+        currentBlockRequest = nullptr;
+    }
+}
+
+bool CBlockRequest::isCancled()
+{
+    return fCancled;
+}
+
+void CBlockRequest::setAsCurrentRequest()
+{
+    // if there is an existing block request, cancle it
+    if (currentBlockRequest != nullptr)
+        currentBlockRequest->fCancled = true;
+
+    currentBlockRequest = shared_from_this();
+}
+
+void CBlockRequest::fillInNextBlocks(std::vector<CBlockIndex*>& vBlocks, unsigned int count, std::function<bool(CBlockIndex*)> filterBlocksCallback)
+{
+    for (unsigned int i = this->processedUpToSize; i < this->vBlocksToDownload.size() ; i++) {
+        CBlockIndex *pindex = this->vBlocksToDownload[i];
+        if ( filterBlocksCallback(pindex) && !(pindex->nStatus & BLOCK_HAVE_DATA)) {
+            // we don't already download this block and we don't have its data already
```

Comment:
this is not quite right English and I don't understand what it is saying. Do you mean "we haven't already downloaded this block nor do we have its data"?